Enforce URL path semantics in containment lookup

NSeydoux opened the pull request:
When listing contained resources with `getContainedResourcesAll`, in addition to using `ldp:contains` claims, we apply a filter so that we only return resources that are child resources of the target container from a URL path semantics perspective as well. This means `[https://pod.example.org/foo/bar/moo`](https://pod.example.org/foo/bar/moo%60) cannot be considered a child resource of `[https://pod.example.org/foo/`](https://pod.example.org/foo/%60). Resources from a different origin will also be excluded.

This prevents malicious Pods from claiming containment across domains, which could lead to incorrect behavior of the clients.

- [X] I've added a unit test to test for potential regressions of this bug.
- [X] The changelog has been updated, if applicable.
- [X] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

NSeydoux committed:
Enforce path semantics in containment lookup

When listing contained resources with `getContainedResourcesAll`, in addition to using `ldp:contains` claims, we apply a filter so that we only return resources that are child resources of the target container from a URL path semantics perspective as well. This means `https://pod.example.org/foo/bar/moo` cannot be considered a child resource of `https://pod.example.org/foo/`. Resources from a different origin will also be excluded.

This prevents malicious Pods from claiming containment across domains, which could lead to incorrect behavior of the clients.

NSeydoux committed:
Updated changelog

jeswr commented on src/resource/solidDataset.ts:
```diff
@@ -844,7 +855,11 @@ export function getContainedResourceUrlAll(
   // > a containment triple MUST be added to the state of the LDPC whose subject is the LDPC URI,
   // > whose predicate is ldp:contains and whose object is the URI for the newly created document
-  return container !== null ? getIriAll(container, ldp.contains) : [];
+  return container !== null
+    ? getIriAll(container, ldp.contains).filter((child) =>
+        isChildResource(child, getSourceUrl(solidDataset))
+      )
+    : [];
```
To me it seems more sensible to do the following (factor out the `getSrcUrl` for performance; and error rather than filtering so that users receive errors to alert them to a malicious Pod rather than results just going missing).

```suggestion
  if (container === null) return [];
  const containment = getIriAll(container, ldp.contains);
  // Make a const outside the loop so we only compute this once
  const srcUrl = getSourceUrl(solidDataset);
  
  for (const child of containment) {
    if (!isChildResource(child, srcUrl))
      // Error rather than just hiding the resource
      throw new Error(`Child resource [${child}] not in container [${srcUrl}] according to slash semantics`);
  }
```

NSeydoux replied:
Agreed, it sounds like throwing here would be a good idea, in particular since there is a direct action the user can take to mitigate the issue: 2e7bebf27d68f188b43ef873cc42bcb874c9c37a

jeswr commented on src/resource/solidDataset.ts:
```diff
@@ -827,6 +827,17 @@ export async function deleteContainer(
+  const urlDiff = child.href.substring(parent.href.length, child.href.length);
+  // The child path component that isn't present in the parent should only
+  // potentially include slashes at the end (if it is a container).
+  const isDirectChild = urlDiff.includes(urlDiff.replace("/", " ").trim());
+  return isAncestor && isDirectChild;
```
The following returns `true` even though I wouldn't expect it to `isChildResource('http://a/b//c/d/e', 'http://a/b/')`.

To fix this I would do something equivalent to
```suggestion
  const stripped = urlDiff.replace(/(^\/)|(\/$)/g, "");
  return isAncestor && stripped.length >= 1 && !stripped.includes("/");
```

jeswr replied:
I'd also be inclined to do this all based off `.path` rather than `.href` and check that all the other `URL` attributes are the same in order to prevent things like `http://example.org/a?q1=a/a` being considered to be contained in `http://example.org/a?q1=a/`

NSeydoux replied:
b55a497ff4775001106d5eceebb1a4b3a770f833 addresses your first point (I didn't realize `.replace` only replaced the first occurrence for non-regexes), and 15dc7f566ece765ff36e87f40ab2bef9762142ee your second point, both very good observations.

jeswr reviewed (changes requested):
Meant to request changes rather than approve

NSeydoux committed:
Detect indirect children on double slash

NSeydoux committed:
Throw on incorrect child resource

NSeydoux committed:
Compute path diff instead of URL diff

This allows to ignore differences in query params.

jeswr commented on src/resource/solidDataset.ts:
```diff
@@ -840,11 +856,24 @@ export async function deleteContainer(
+    // See https://solidproject.org/TR/protocol#resource-containment
+    if (!isChildResource(childUrl, containerUrl)) {
+      throw new Error(
+        `Container [${containerUrl}] claims containment of [${childUrl}], but this is not possible according to slash semantics. The containment triple should be removed from [${containerUrl}].`
```
```suggestion
        `Container [${containerUrl}] claims containment of [${childUrl}], but this is not possible according to slash semantics (https://solidproject.org/TR/protocol#resource-containment). The containment triple should be removed from [${containerUrl}].`
```

jeswr commented on src/resource/solidDataset.ts:
```diff
@@ -827,6 +827,22 @@ export async function deleteContainer(
+  // The child path component that isn't present in the parent should only
+  // potentially include slashes at the end (if it is a container).
+  const isDirectChild =
+    urlDiff !== "" && urlDiff.includes(urlDiff.replace(/\//g, " ").trim());
```
I still find this less intuitive and more buggy than https://github.com/inrupt/solid-client-js/pull/2059#discussion_r1248579674.

In particular; it looks like `isDirectChild` will incorrectly return true when `urlDiff` is the following values `/`, `//`, `///` etc; `/ /a//`. You could even do

```ts
if (urlDiff.startsWith('/')) urlDiff = urlDiff.slice(1)
if (urlDiff.endsWith('/')) urlDiff = urlDiff.slice(0, -1)
isDirectChild = urlDiff !== "" && !urlDiff.includes("/")
```

Secondly should we check to make sure that `search` attribute of both `URL`s is the empty string.

NSeydoux replied:
You're right, it is cleaner, implemented in 3b5aa3c26786c23715c614c509c303fb08ad98f7.

To your second point, I don't know if the specification says anything about search params. There probably won't be search params in most cases, but if the spec doesn't explicitly prevent it let's not make assumptions.

NSeydoux committed:
Update src/resource/solidDataset.ts

Co-authored-by: Jesse Wright <63333554+jeswr@users.noreply.github.com>

NSeydoux committed:
Clarify path semantics enforcement

NSeydoux commented:
Note that there is a discussion in https://github.com/inrupt/solid-client-java/pull/571 about whether we should throw or not, I'd like to have consistency across SDKs for such behaviors.

NSeydoux committed:
Filter instead of throw

Instead of throwing, getContainedResourceAll filters out invalid resources. In addition, a validation function allows to check whether there are inalid containment triples.

NSeydoux committed:
Export new function

NSeydoux committed:
Update changelog

NSeydoux committed:
Complete code coverage

jeswr committed:
Merge branch 'main' into fix/SDK-3153_path-semantics-containment
